service-netty/src/main/java/info/bitrich/xchangestream/service/netty/NettyStreamingService.java

```diff
@@ -225,6 +227,10 @@ public void sendMessage(String message) {
         }
     }
 
+    public Observable<Throwable> subscribeReconnectFailure() {
+        return Observable.<Throwable>create(observableEmitter -> reconnFailEmitters.add(observableEmitter));
```

Comment:
Probably more reactive-standard here to have a single `Flowable` to which we emit the reconnect failures, set up with `.share()`. This method would then just subscribe to that as an `Observable`.

I'm no Rx expert, but am thinking something like this:

```
private final AtomicReference<FlowableEmitter<Throwable>> emitter = new AtomicReference<>();
private final Flowable<Throwable> reconnectFailures = Flowable.create(
    (FlowableEmitter<Throwable> e) -> emitter.set(e.serialize()),
    BackpressureStrategy.MISSING
  ).share();
```

Then:

```
public Observable<Throwable> subscribeReconnectFailure() {
  return reconnectFailures;
}
```

And:

```
 .doOnError(t -> {
  LOG.warn("Problem with reconnect", t);
  emitter.onNext(t);
})
```

Reply:
Thank you four your advice, which is great.  I had no experience of Rx at all so the previous implementation is not good. I have modified following your suggestions.